zo_compression_main.py

```diff
@@ -0,0 +1,217 @@
+        optimizer = torch.optim.SGD(model.parameters(), lr=args.lr, momentum=0, weight_decay=5e-4)
+        accuracy_func = get_lm_loss("accuracy", verbalizer_id_map)
+    else:
+        raise Exception(f"Dataset {args.dataset} is not supported")
+
+    if args.grad_estimate_method in ["rge-central", "rge-forward"]:
+        method = args.grad_estimate_method[4:]
+        print(f"Using RGE {method}")
+        grad_estimator = RGE(
+            model,
+            mu=args.mu,
+            num_pert=args.num_pert,
+            grad_estimate_method=method,
+            device=device,
+            torch_dtype=torch_dtype,
+        )
+    else:
+        raise Exception(f"Grad estimate method {args.grad_estimate_method} not supported")
+    return model, criterion, optimizer, grad_estimator, accuracy_func
+
+
+def setup_server_and_clients(
+    args, device_map: dict[str, torch.device], train_loaders
+) -> CeZO_Server:
+    clients = []
+
+    for i in range(args.num_clients):
+        client_name = get_client_name(i)
+        client_device = device_map[client_name]
+        (
+            client_model,
+            client_criterion,
+            client_optimizer,
+            client_grad_estimator,
+            client_accuracy_func,
+        ) = prepare_settings_underseed(args, client_device)
+        client_model.to(client_device)
+
+        client = ResetClient(
+            client_model,
+            train_loaders[i],
+            client_grad_estimator,
+            client_optimizer,
+            client_criterion,
+            client_accuracy_func,
+            client_device,
+        )
+        clients.append(client)
+
+    server_device = device_map["server"]
+    server = CeZO_Server(
+        clients,
+        server_device,
+        num_sample_clients=args.num_sample_clients,
+        local_update_steps=args.local_update_steps,
+    )
+
+    # set server tools
+    (
+        server_model,
+        server_criterion,
+        server_optimizer,
+        server_grad_estimator,
+        server_accuracy_func,
+    ) = prepare_settings_underseed(args, server_device)
+    server_model.to(server_device)
+    server.set_server_model_and_criterion(
+        server_model,
+        server_criterion,
+        server_accuracy_func,
+        server_optimizer,
+        server_grad_estimator,
+    )
+    return server
+
+
+# get_warmup_lr is not used for now.
+def get_warmup_lr(args, current_epoch: int, current_iter: int, iters_per_epoch: int) -> float:
+    overall_iterations = args.warmup_epochs * iters_per_epoch + 1
+    current_iterations = current_epoch * iters_per_epoch + current_iter + 1
+    return args.lr * current_iterations / overall_iterations
+
+
+def get_size_of_model(model):
+    return sum(p.numel() * p.element_size() for p in model.parameters())
+
+
+if __name__ == "__main__":
+    args = get_params().parse_args()
+    if args.dataset == "shakespeare":
+        args.num_clients = 139
```

Comment:
delete here as well

Reply:
Deleted.

Reply:
If I keep the Shakespeare dataset, should I keep these codes?

Reply:
See my comment above. Still prefer deleting it.